identity-iota/src/did/doc/iota_document.rs

```diff
@@ -99,19 +102,65 @@ impl IotaDocument {
       .authentication(method)
       .build()
       .map(CoreDocument::into_verifiable)
-      .map(Into::into)
+      .map(TryInto::try_into)
+      .map(Result::unwrap)
       .unwrap() // `unwrap` is fine - we provided all the necessary properties
   }
 
-  /// Converts a generic DID [`Document`][CoreDocument] to an IOTA DID Document.
+  /// Converts a generic DID [`Document`][`CoreDocument`] to an IOTA DID Document.
   ///
   /// # Errors
   ///
   /// Returns `Err` if the document is not a valid IOTA DID Document.
   pub fn try_from_core(document: CoreDocument) -> Result<Self> {
+    IotaDocument::validate_core_document(&document)?;
+
+    Ok(Self {
+      document: document.serde_into()?,
+      message_id: MessageId::null(),
+    })
+  }
+
+  /// Converts a generic DID [`Document`][`BaseDocument`] to an IOTA DID Document.
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the document is not a valid IOTA DID Document.
+  pub fn try_from_base(document: BaseDocument) -> Result<Self> {
+    IotaDocument::validate_core_document(&document)?;
+
+    Ok(Self {
+      document: document.serde_into()?,
+      message_id: MessageId::null(),
+    })
+  }
+
+  /// Performs validation that a `CoreDocument` adheres to the IOTA spec.
+  ///
+  /// # Errors
+  ///
+  /// Returns `Err` if the document is not a valid IOTA DID Document.
+  fn validate_core_document<T, U, V>(document: &CoreDocument<T, U, V>) -> Result<()> {
+    // Validate that the DID conforms to the IotaDID specification.
+    // This check is required to ensure the correctness of the `IotaDocument::id()` method which
+    // creates an `IotaDID::new_unchecked_ref()` from the underlying DID.
     let did: &IotaDID = IotaDID::try_from_borrowed(document.id())?;
 
-    let method: &VerificationMethod = document
+    // Validate that the document controller (if any) conforms to the IotaDID specification.
+    // This check is required to ensure the correctness of the `IotaDocument::controller()` method which
+    // creates an `IotaDID::new_unchecked_ref()` from the underlying controller.
+    document.controller().map_or(Ok(()), |c| IotaDID::check_validity(c))?;
+
+    // Validate that the validation methods conform to the IotaDID specification.
+    // This check is required to ensure the correctness of the `IotaDocument::methods()`,
+    // `IotaDocument::resolve()`, `IotaDocument::try_resolve()`, IotaDocument::resolve_mut()`,
+    // and IotaDocument::try_resolve_mut()` methods which creates an `IotaDID::new_unchecked_ref()`
+    // from the underlying controller.
+    for method in document.methods() {
```

Comment:
This only checks the embedded verification methods and not ones referenced by `DID`.

Currently the only way to **add** a non-IOTA DID verification method is by converting to/from JSON but I think a simple fix would be to add `Document::method_refs` that returns the unfiltered `MethodRef` type.

You could then `match` in the iterator and either call `IotaVerificationMethod::check_validity` or `IotaDID::check_validity`

Reply:
Ahh good catch, I glossed over the `__filter_ref` in `CoreDocument::methods()`.

So, the one tricky thing is the `self.verification_method` part of that iterator `.chain()` in `CoreDocument::methods()` because this is a `VerificationMethod<T>`, not a `MethodRef<T>` (like all the others). This means you can't construct a `MethodRef::Embed` from a reference, so (afaik) we'd need to `Clone` the underlying `VerificationMethod` to construct the iterator, which seems non-ideal from a performance standpoint. lmk if I'm missing something though.

Reply:
> This means you can't construct a MethodRef::Embed from a reference

😵 I glossed over this as well. you're right that `Clone` would be non-ideal.

Maybe this is better suited as just an iterator over the [verification relationships](https://www.w3.org/TR/2020/WD-did-core-20200731/#verification-relationships) and you could use that in combination with `document.verification_method()`.

```
for method_ref in document.verification_relationships() {
```

Reply:
Seems resonable, done!